File: internal/csvons/field_occurrences.go
package csvons

import (
	"log"
	"slices"
	"strings"
)

// FieldOccurrence captures a value emitted from a field expression together
// with the 1-based CSV row number it came from.
type FieldOccurrence struct {
	Row   int
	Value string
}

type fieldOccurrenceProvider interface {
	FieldOccurrences(fields []string, records [][]string) <-chan FieldOccurrence
}

func (p *PlainField) FieldOccurrences(fields []string, records [][]string) <-chan FieldOccurrence {
	fieldIndex := slices.Index(fields, p.fieldName)
	if fieldIndex == -1 {
		return nil
	}

	output := make(chan FieldOccurrence, 128)
	go func() {
		defer close(output)
		for i := p.metadata.DataIndex; i < len(records); i++ {
			record := records[i]
			if fieldIndex < len(record) {
				output <- FieldOccurrence{Row: i + 1, Value: record[fieldIndex]}
			} else {
				log.Printf("plain field [%s] not found in record [%d]", p.fieldName, i)
			}
		}
	}()

	return output
}

func (r *RepeatField) FieldOccurrences(fields []string, records [][]string) <-chan FieldOccurrence {
	fieldIndex := slices.Index(fields, r.fieldName)
	if fieldIndex == -1 {
		return nil
	}

	output := make(chan FieldOccurrence, 128)
	go func() {
		defer close(output)
		for i := r.metadata.DataIndex; i < len(records); i++ {
			record := records[i]
			if fieldIndex < len(record) {
				lev1Vals := strings.Split(record[fieldIndex], r.metadata.Lev1Separator)
				for _, lev1Val := range lev1Vals {
					output <- FieldOccurrence{Row: i + 1, Value: lev1Val}
				}
			} else {
				log.Printf("repeat field [%s] not found in record [%d]", r.fieldName, i)
			}
		}
	}()

	return output
}

func (n *NestedField) FieldOccurrences(fields []string, records [][]string) <-chan FieldOccurrence {
	fieldIndex := slices.Index(fields, n.fieldName)
	if fieldIndex == -1 {
		return nil
	}

	output := make(chan FieldOccurrence, 128)
	go func() {
		defer close(output)
		for i := n.metadata.DataIndex; i < len(records); i++ {
			record := records[i]
			if fieldIndex < len(record) {
				lev1Vals := strings.Split(record[fieldIndex], n.metadata.Lev1Separator)
				for _, lev1Val := range lev1Vals {
					lev2Vals := strings.Split(lev1Val, n.metadata.Lev2Separator)
					if n.index < len(lev2Vals) {
						output <- FieldOccurrence{Row: i + 1, Value: lev2Vals[n.index]}
					} else {
						log.Printf("nested field [%s] level 2 value [%s] length [%d] not found in record [%d]", n.fieldName, lev1Val, len(lev2Vals), i)
					}
				}
			} else {
				log.Printf("nested field [%s] not found in record [%d]", n.fieldName, i)
			}
		}
	}()

	return output
}

func (c *ComplexField) FieldOccurrences(fields []string, records [][]string) <-chan FieldOccurrence {
	fieldIndexes := make([]int, len(c.fieldNames))
	for i, fieldName := range c.fieldNames {
		fieldIndexes[i] = slices.Index(fields, fieldName)
		if fieldIndexes[i] == -1 {
			log.Printf("complex field [%s] not found in fields", fieldName)
			return nil
		}
	}

	output := make(chan FieldOccurrence, 128)
	go func() {
		defer close(output)
		for i := c.metadata.DataIndex; i < len(records); i++ {
			record := records[i]
			cpxStr := ""
			for idx, fieldIndex := range fieldIndexes {
				if fieldIndex < len(record) {
					cpxStr += record[fieldIndex] + c.metadata.FieldConnector
				} else {
					log.Printf("complex field [%s] not found in record [%d]", c.fieldNames[idx], i)
					return
				}
			}
			output <- FieldOccurrence{Row: i + 1, Value: cpxStr}
		}
	}()

	return output
}

File: gui/electron/src/core/field-occurrences.ts
import type { Metadata } from './types.ts';
import type { FieldExpr } from './field-expr.ts';
import {
  PlainField,
  RepeatField,
  NestedField,
  ComplexField,
} from './field-expr.ts';

/**
 * FieldOccurrence captures a value with its 1-based CSV row number.
 */
export interface FieldOccurrence {
  row: number;
  value: string;
}

/**
 * FieldOccurrenceProvider can yield field occurrences with row tracking.
 */
export interface FieldOccurrenceProvider extends FieldExpr {
  fieldOccurrences(
    fields: string[],
    records: string[][],
  ): Iterable<FieldOccurrence> | null;
}

export function isFieldOccurrenceProvider(
  expr: FieldExpr,
): expr is FieldOccurrenceProvider {
  return 'fieldOccurrences' in expr;
}

// Extend PlainField with FieldOccurrences
PlainField.prototype.fieldOccurrences = function* (
  this: PlainField,
  fields: string[],
  records: string[][],
): Generator<FieldOccurrence> {
  const fieldIndex = fields.indexOf(this.fieldName);
  if (fieldIndex === -1) return;

  for (let i = this.metadata.data_index; i < records.length; i++) {
    const record = records[i]!;
    if (fieldIndex < record.length) {
      yield { row: i + 1, value: record[fieldIndex]! };
    }
  }
};

// Extend RepeatField with FieldOccurrences
RepeatField.prototype.fieldOccurrences = function* (
  this: RepeatField,
  fields: string[],
  records: string[][],
): Generator<FieldOccurrence> {
  const fieldIndex = fields.indexOf(this.fieldName);
  if (fieldIndex === -1) return;

  for (let i = this.metadata.data_index; i < records.length; i++) {
    const record = records[i]!;
    if (fieldIndex < record.length) {
      const lev1Vals = record[fieldIndex]!.split(this.metadata.lev1_separator);
      for (const val of lev1Vals) {
        yield { row: i + 1, value: val };
      }
    }
  }
};

// Extend NestedField with FieldOccurrences
NestedField.prototype.fieldOccurrences = function* (
  this: NestedField,
  fields: string[],
  records: string[][],
): Generator<FieldOccurrence> {
  const fieldIndex = fields.indexOf(this.fieldName);
  if (fieldIndex === -1) return;

  for (let i = this.metadata.data_index; i < records.length; i++) {
    const record = records[i]!;
    if (fieldIndex < record.length) {
      const lev1Vals = record[fieldIndex]!.split(this.metadata.lev1_separator);
      for (const lev1Val of lev1Vals) {
        const lev2Vals = lev1Val.split(this.metadata.lev2_separator);
        if (this.index < lev2Vals.length) {
          yield { row: i + 1, value: lev2Vals[this.index]! };
        }
      }
    }
  }
};

// Extend ComplexField with FieldOccurrences
ComplexField.prototype.fieldOccurrences = function* (
  this: ComplexField,
  fields: string[],
  records: string[][],
): Generator<FieldOccurrence> {
  const fieldIndexes = this.fieldNames.map((name) => fields.indexOf(name));
  if (fieldIndexes.some((idx) => idx === -1)) return;

  for (let i = this.metadata.data_index; i < records.length; i++) {
    const record = records[i]!;
    let cpxStr = '';
    let valid = true;
    for (const fieldIndex of fieldIndexes) {
      if (fieldIndex < record.length) {
        cpxStr += record[fieldIndex]! + this.metadata.field_connector;
      } else {
        valid = false;
        break;
      }
    }
    if (valid) {
      yield { row: i + 1, value: cpxStr };
    }
  }
};

// Augment the class interfaces for TypeScript
declare module './field-expr.ts' {
  interface PlainField extends FieldOccurrenceProvider {}
  interface RepeatField extends FieldOccurrenceProvider {}
  interface NestedField extends FieldOccurrenceProvider {}
  interface ComplexField extends FieldOccurrenceProvider {}
}
